File: windows/Services/MsuOriginalSoundtrack.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Net.Http;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using LTTPRandomizerGenerator.Models;

namespace LTTPRandomizerGenerator.Services;

public static partial class MsuOriginalSoundtrack
{
    private static readonly string CacheDir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "LTTPRandomizerGenerator", "OriginalAudio");

    private static readonly HashSet<string> SupportedExtensions =
        new(StringComparer.OrdinalIgnoreCase) { ".mp3", ".wav", ".wma", ".aac", ".m4a", ".aiff", ".aif", ".flac" };

    private static readonly Dictionary<string, int> OstAliases = new(StringComparer.OrdinalIgnoreCase)
    {
        ["title"]                    = 1,
        ["link to the past"]         = 1,
        ["beginning of the journey"] = 6,
        ["seal of seven maidens"]    = 3,
        ["time of the falling rain"] = 3,
        ["majestic castle"]          = 16,
        ["princess zeldas rescue"]   = 25,
        ["safety in the sanctuary"]  = 20,
        ["hyrule field main theme"]  = 2,
        ["hyrule field"]             = 2,
        ["kakariko village"]         = 7,
        ["guessing game house"]      = 14,
        ["fortune teller"]           = 23,
        ["soldiers of kakariko"]     = 12,
        ["dank dungeons"]            = 17,
        ["lost ancient ruins"]       = 18,
        ["anger of the guardians"]   = 21,
        ["great victory"]            = 19,
        ["silly pink rabbit"]        = 4,
        ["forest of mystery"]        = 5,
        ["unsealing the master sword"] = 10,
        ["priest of the dark order"] = 28,
        ["dark golden land"]         = 9,
        ["black mist"]               = 13,
        ["dungeon of shadows"]       = 22,
        ["meeting the maidens"]      = 26,
        ["goddess appears"]          = 27,
        ["release of ganon"]         = 29,
        ["ganons message"]           = 30,
        ["prince of darkness"]       = 31,
        ["power of the gods"]        = 32,
        ["epilogue"]                 = 33,
        ["beautiful hyrule"]         = 33,
        ["staff roll"]               = 34,
        ["credits"]                  = 34,
        ["overworld"]                = 2,
        ["dark world theme"]         = 9,
        ["file select"]              = 11,
        ["game over"]                = 11,
        ["sanctuary"]                = 20,
        ["lost woods"]               = 5,
        ["kakariko"]                 = 7,
        ["ganon battle"]             = 31,
        ["ganon fight"]              = 31,
        ["triforce"]                 = 32,
        ["fairy fountain"]           = 27,
        ["boss battle"]              = 21,
        ["boss victory"]             = 19,
        ["cave"]                     = 18,
        ["shop"]                     = 23,
        ["minigame"]                 = 14,
        ["skull woods"]              = 15,
        ["dimensional shift"]        = 8,
        ["teleport"]                 = 8,
    };

    public static void LoadCachedOriginals(IReadOnlyList<MsuTrackSlot> tracks)
    {
        if (!Directory.Exists(CacheDir)) return;

        foreach (var track in tracks)
        {
            string cached = Path.Combine(CacheDir, $"{track.SlotNumber:D2}.pcm");
            if (File.Exists(cached))
                track.OriginalPcmPath = cached;
        }
    }

    public static void ClearCache(IReadOnlyList<MsuTrackSlot> tracks)
    {
        foreach (var track in tracks)
            track.OriginalPcmPath = null;

        if (!Directory.Exists(CacheDir)) return;
        try { Directory.Delete(CacheDir, true); } catch { }
    }

    public static async Task<string?> ImportFromFolderAsync(
        string folderPath,
        IReadOnlyList<MsuTrackSlot> tracks,
        IProgress<(int current, int total, string trackName)>? progress = null,
        CancellationToken ct = default)
    {
        if (!Directory.Exists(folderPath))
            return "Folder does not exist.";

        var audioFiles = Directory.GetFiles(folderPath)
            .Where(f => SupportedExtensions.Contains(Path.GetExtension(f)))
            .ToList();

        if (audioFiles.Count == 0)
            return "No supported audio files found. Supported: MP3, WAV, WMA, AAC, M4A, AIFF, FLAC.";

        var matches = MatchFilesToSlots(audioFiles, tracks);
        if (matches.Count == 0)
            return "Could not match any files to track slots. Files should be numbered (e.g., '01 - Opening.mp3') or named to match track names.";

        return await ConvertAndCacheAsync(matches, tracks, progress, ct);
    }

    public static async Task<string?> ImportFromZipAsync(
        string zipPath,
        IReadOnlyList<MsuTrackSlot> tracks,
        IProgress<(int current, int total, string trackName)>? progress = null,
        CancellationToken ct = default)
    {
        if (!File.Exists(zipPath))
            return "ZIP file does not exist.";

        string tempDir = Path.Combine(Path.GetTempPath(), "LTTPOriginalImport_" + Guid.NewGuid().ToString("N")[..8]);

        try
        {
            Directory.CreateDirectory(tempDir);
            await Task.Run(() => ZipFile.ExtractToDirectory(zipPath, tempDir), ct);

            var audioFiles = Directory.GetFiles(tempDir, "*.*", SearchOption.AllDirectories)
                .Where(f => SupportedExtensions.Contains(Path.GetExtension(f)))
                .ToList();

            if (audioFiles.Count == 0)
                return "No supported audio files found in ZIP.";

            var matches = MatchFilesToSlots(audioFiles, tracks);
            if (matches.Count == 0)
                return "Could not match any files to track slots.";

            return await ConvertAndCacheAsync(matches, tracks, progress, ct);
        }
        finally
        {
            try { Directory.Delete(tempDir, true); } catch { }
        }
    }

    public static async Task<string?> ImportFromUrlAsync(
        string url,
        IReadOnlyList<MsuTrackSlot> tracks,
        HttpClient http,
        IProgress<(int current, int total, string trackName)>? progress = null,
        CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(url))
            return "URL cannot be empty.";

        string tempZip = Path.Combine(Path.GetTempPath(), "LTTPOriginal_" + Guid.NewGuid().ToString("N")[..8] + ".zip");

        try
        {
            progress?.Report((0, 1, "Downloading…"));

            using var response = await http.GetAsync(url, HttpCompletionOption.ResponseHeadersRead, ct);
            if (!response.IsSuccessStatusCode)
                return $"Download failed: HTTP {(int)response.StatusCode} {response.ReasonPhrase}";

            await using (var stream = await response.Content.ReadAsStreamAsync(ct))
            await using (var file = File.Create(tempZip))
            {
                await stream.CopyToAsync(file, ct);
            }

            return await ImportFromZipAsync(tempZip, tracks, progress, ct);
        }
        catch (OperationCanceledException) { throw; }
        catch (Exception ex)
        {
            return $"Download failed: {ex.Message}";
        }
        finally
        {
            try { File.Delete(tempZip); } catch { }
        }
    }

    public static Dictionary<int, string> MatchFilesToSlots(
        IReadOnlyList<string> filePaths,
        IReadOnlyList<MsuTrackSlot> tracks)
    {
        var result = new Dictionary<int, string>();
        var slotLookup = tracks.ToDictionary(t => t.SlotNumber);
        var usedSlots = new HashSet<int>();

        // Pass 1: OST alias table
        foreach (var filePath in filePaths)
        {
            string fileName = Path.GetFileNameWithoutExtension(filePath);
            string cleaned = LeadingNumberRegex().Replace(fileName, "")
                .Replace('_', ' ').Replace('-', ' ').Replace('~', ' ')
                .Replace(".", " ").Replace("'", "").Replace("\u2019", "").Trim();

            foreach (var (alias, slot) in OstAliases)
            {
                if (usedSlots.Contains(slot)) continue;
                if (!slotLookup.ContainsKey(slot)) continue;

                if (cleaned.Contains(alias, StringComparison.OrdinalIgnoreCase))
                {
                    result[slot] = filePath;
                    usedSlots.Add(slot);
                    break;
                }
            }
        }

        // Pass 2: leading number
        foreach (var filePath in filePaths)
        {
            if (result.ContainsValue(filePath)) continue;

            string fileName = Path.GetFileNameWithoutExtension(filePath);
            var match = LeadingNumberRegex().Match(fileName);
            if (match.Success && int.TryParse(match.Value, out int slot) && slot >= 1 && slot <= 61 && slotLookup.ContainsKey(slot))
            {
                if (usedSlots.Add(slot))
                    result[slot] = filePath;
            }
        }

        // Pass 3: any number
        foreach (var filePath in filePaths)
        {
            if (result.ContainsValue(filePath)) continue;

            string fileName = Path.GetFileNameWithoutExtension(filePath);
            var matches = AnyNumberRegex().Matches(fileName);
            foreach (Match m in matches)
            {
                if (int.TryParse(m.Value, out int slot) && slot >= 1 && slot <= 61 && slotLookup.ContainsKey(slot) && usedSlots.Add(slot))
                {
                    result[slot] = filePath;
                    break;
                }
            }
        }

        // Pass 4: fuzzy name match
        foreach (var filePath in filePaths)
        {
            if (result.ContainsValue(filePath)) continue;

            string fileName = Path.GetFileNameWithoutExtension(filePath)
                .Replace('_', ' ').Replace('-', ' ').Trim();

            foreach (var track in tracks)
            {
                if (usedSlots.Contains(track.SlotNumber)) continue;

                if (fileName.Contains(track.Name, StringComparison.OrdinalIgnoreCase) ||
                    track.Name.Contains(fileName, StringComparison.OrdinalIgnoreCase))
                {
                    result[track.SlotNumber] = filePath;
                    usedSlots.Add(track.SlotNumber);
                    break;
                }
            }
        }

        return result;
    }

    private static async Task<string?> ConvertAndCacheAsync(
        Dictionary<int, string> matches,
        IReadOnlyList<MsuTrackSlot> tracks,
        IProgress<(int current, int total, string trackName)>? progress,
        CancellationToken ct)
    {
        Directory.CreateDirectory(CacheDir);

        var slotLookup = tracks.ToDictionary(t => t.SlotNumber);
        var sorted = matches.OrderBy(kv => kv.Key).ToList();
        int completed = 0;
        int failed = 0;

        foreach (var (slot, sourcePath) in sorted)
        {
            ct.ThrowIfCancellationRequested();

            string trackName = slotLookup.TryGetValue(slot, out var ts) ? ts.Name : $"Track {slot}";
            progress?.Report((completed + 1, sorted.Count, trackName));

            string destPath = Path.Combine(CacheDir, $"{slot:D2}.pcm");

            if (!File.Exists(destPath))
            {
                string? error = await MsuPcmConverter.ConvertAsync(sourcePath, destPath, ct: ct);
                if (error is not null)
                {
                    failed++;
                    completed++;
                    continue;
                }
            }

            if (slotLookup.TryGetValue(slot, out var trackSlot))
                trackSlot.OriginalPcmPath = destPath;

            completed++;
        }

        if (failed > 0 && failed == sorted.Count)
            return "All conversions failed. Ensure the files are valid audio files.";

        if (failed > 0)
            return $"{sorted.Count - failed} of {sorted.Count} tracks imported. {failed} file(s) failed to convert.";

        return null;
    }

    [GeneratedRegex(@"^\d+")]
    private static partial Regex LeadingNumberRegex();

    [GeneratedRegex(@"\d+")]
    private static partial Regex AnyNumberRegex();
}

File: android/app/src/main/java/com/lttprandomizer/MsuOriginalSoundtrack.kt
package com.lttprandomizer

import android.content.Context
import java.io.File
import java.util.zip.ZipInputStream

object MsuOriginalSoundtrack {
    private fun cacheDir(context: Context) = File(context.filesDir, "original_audio")

    private val audioExtensions = setOf("pcm", "mp3", "ogg", "wav", "flac", "m4a", "aac")

    private fun isAudioFile(name: String): Boolean {
        val ext = name.substringAfterLast('.', "").lowercase()
        return ext in audioExtensions
    }

    // Sorted longest-first so specific aliases match before generic substrings.
    // e.g. "soldiers of kakariko" must match before "kakariko village" or "kakariko".
    private val ostAliases: List<Pair<String, Int>> = listOf(
        "soldiers of kakariko" to 12,
        "beginning of the journey" to 6,
        "unsealing the master sword" to 10,
        "hyrule field main theme" to 2,
        "safety in the sanctuary" to 20,
        "priest of the dark order" to 28,
        "time of the falling rain" to 3,
        "seal of seven maidens" to 3,
        "anger of the guardians" to 21,
        "princess zeldas rescue" to 25,
        "meeting the maidens" to 26,
        "prince of darkness" to 31,
        "dungeon of shadows" to 22,
        "guessing game house" to 14,
        "kakariko village" to 7,
        "lost ancient ruins" to 18,
        "dark golden land" to 9,
        "dimensional shift" to 8,
        "release of ganon" to 29,
        "majestic castle" to 16,
        "forest of mystery" to 5,
        "silly pink rabbit" to 4,
        "power of the gods" to 32,
        "dark world theme" to 9,
        "goddess appears" to 27,
        "ganons message" to 30,
        "link to the past" to 1,
        "beautiful hyrule" to 33,
        "fairy fountain" to 27,
        "dank dungeons" to 17,
        "great victory" to 19,
        "hyrule field" to 2,
        "fortune teller" to 23,
        "boss victory" to 19,
        "black mist" to 13,
        "skull woods" to 15,
        "ganon battle" to 31,
        "boss battle" to 21,
        "ganon fight" to 31,
        "dark world" to 9,
        "lost woods" to 5,
        "staff roll" to 34,
        "quit game" to 11,
        "game over" to 11,
        "sanctuary" to 20,
        "overworld" to 2,
        "kakariko" to 7,
        "teleport" to 8,
        "triforce" to 32,
        "epilogue" to 33,
        "minigame" to 14,
        "credits" to 34,
        "title" to 1,
        "cave" to 18,
        "shop" to 23,
    )

    fun loadCachedOriginals(context: Context, tracks: List<MsuTrackSlot>) {
        val dir = cacheDir(context)
        if (!dir.exists()) return

        val cachedFiles = dir.listFiles() ?: return
        for (track in tracks) {
            val match = cachedFiles.firstOrNull {
                it.nameWithoutExtension == track.slotDisplay && isAudioFile(it.name)
            }
            if (match != null) track.originalPcmPath = match.absolutePath
        }
    }

    fun clearCache(context: Context, tracks: List<MsuTrackSlot>) {
        for (track in tracks) track.originalPcmPath = null
        val dir = cacheDir(context)
        if (dir.exists()) dir.deleteRecursively()
    }

    /** Import audio files from a folder. Supports PCM, MP3, OGG, WAV, FLAC, M4A, AAC. */
    fun importFromFolder(context: Context, folderPath: String, tracks: List<MsuTrackSlot>): String? {
        val folder = File(folderPath)
        if (!folder.exists()) return "Folder does not exist."

        val audioFiles = folder.listFiles()?.filter { isAudioFile(it.name) } ?: emptyList()
        if (audioFiles.isEmpty()) return "No audio files found in folder."

        val matches = matchFilesToSlots(audioFiles.map { it.absolutePath }, tracks)
        if (matches.isEmpty()) return "Could not match any files to track slots."

        return copyToCache(context, matches, tracks)
    }

    /** Import audio files from a ZIP. Supports PCM, MP3, OGG, WAV, FLAC, M4A, AAC. */
    fun importFromZip(context: Context, inputStream: java.io.InputStream, tracks: List<MsuTrackSlot>): String? {
        val tempDir = File(context.cacheDir, "ost_import_${System.currentTimeMillis()}")
        tempDir.mkdirs()

        try {
            val zipStream = ZipInputStream(inputStream)
            var entry = zipStream.nextEntry
            while (entry != null) {
                if (!entry.isDirectory && isAudioFile(entry.name)) {
                    val destFile = File(tempDir, entry.name.substringAfterLast('/'))
                    if (destFile.canonicalPath.startsWith(tempDir.canonicalPath)) {
                        destFile.outputStream().use { out -> zipStream.copyTo(out) }
                    }
                }
                zipStream.closeEntry()
                entry = zipStream.nextEntry
            }
            zipStream.close()

            val audioFiles = tempDir.listFiles()?.filter { isAudioFile(it.name) } ?: emptyList()
            if (audioFiles.isEmpty()) return "No audio files found in ZIP."

            val matches = matchFilesToSlots(audioFiles.map { it.absolutePath }, tracks)
            if (matches.isEmpty()) return "Could not match any files to track slots."

            return copyToCache(context, matches, tracks)
        } finally {
            tempDir.deleteRecursively()
        }
    }

    fun matchFilesToSlots(filePaths: List<String>, tracks: List<MsuTrackSlot>): Map<Int, String> {
        val result = mutableMapOf<Int, String>()
        val slotLookup = tracks.associateBy { it.slotNumber }
        val usedSlots = mutableSetOf<Int>()
        val aliasMatchedFiles = mutableSetOf<String>() // files that matched an alias (even if slot taken)
        val leadingNumberRegex = Regex("^\\d+")
        val anyNumberRegex = Regex("\\d+")

        // Sort: non-variant files first so "Majestic Castle" beats "Majestic Castle (Storm)"
        val variantPattern = Regex("""\(storm\)|\(short\)|\(variant\)""", RegexOption.IGNORE_CASE)
        val sortedPaths = filePaths.sortedBy { if (variantPattern.containsMatchIn(File(it).nameWithoutExtension)) 1 else 0 }

        // Pass 1: alias table
        for (filePath in sortedPaths) {
            val fileName = File(filePath).nameWithoutExtension
            val cleaned = leadingNumberRegex.replace(fileName, "")
                .replace('_', ' ').replace('-', ' ').replace('~', ' ')
                .replace(".", " ").replace("'", "").replace("\u2019", "").trim()

            for ((alias, slot) in ostAliases) {
                if (slot !in slotLookup) continue
                if (cleaned.contains(alias, ignoreCase = true)) {
                    aliasMatchedFiles.add(filePath)
                    if (slot !in usedSlots) {
                        result[slot] = filePath
                        usedSlots.add(slot)
                    }
                    break // first matching alias wins (list is sorted longest-first)
                }
            }
        }

        // Pass 2: leading number (skip files that matched an alias — their number is likely a disc track, not a slot)
        for (filePath in sortedPaths) {
            if (filePath in result.values || filePath in aliasMatchedFiles) continue
            val fileName = File(filePath).nameWithoutExtension
            val match = leadingNumberRegex.find(fileName) ?: continue
            val slot = match.value.toIntOrNull() ?: continue
            if (slot in 1..61 && slot in slotLookup && usedSlots.add(slot))
                result[slot] = filePath
        }

        // Pass 3: any number (skip alias-matched files)
        for (filePath in sortedPaths) {
            if (filePath in result.values || filePath in aliasMatchedFiles) continue
            val fileName = File(filePath).nameWithoutExtension
            for (match in anyNumberRegex.findAll(fileName)) {
                val slot = match.value.toIntOrNull() ?: continue
                if (slot in 1..61 && slot in slotLookup && usedSlots.add(slot)) {
                    result[slot] = filePath
                    break
                }
            }
        }

        // Pass 4: fuzzy name match
        for (filePath in sortedPaths) {
            if (filePath in result.values) continue
            val fileName = File(filePath).nameWithoutExtension.replace('_', ' ').replace('-', ' ').trim()
            for (track in tracks) {
                if (track.slotNumber in usedSlots) continue
                if (fileName.contains(track.name, ignoreCase = true) ||
                    track.name.contains(fileName, ignoreCase = true)) {
                    result[track.slotNumber] = filePath
                    usedSlots.add(track.slotNumber)
                    break
                }
            }
        }

        return result
    }

    private fun copyToCache(context: Context, matches: Map<Int, String>, tracks: List<MsuTrackSlot>): String? {
        val dir = cacheDir(context)
        dir.mkdirs()

        val slotLookup = tracks.associateBy { it.slotNumber }
        var failed = 0

        for ((slot, sourcePath) in matches.toSortedMap()) {
            val ext = File(sourcePath).extension.ifEmpty { "pcm" }
            val slotName = slot.toString().padStart(2, '0')
            // Remove any existing file for this slot (might have different extension)
            dir.listFiles()?.filter { it.nameWithoutExtension == slotName }?.forEach { it.delete() }
            val destPath = File(dir, "$slotName.$ext")
            try {
                File(sourcePath).copyTo(destPath, overwrite = true)
                slotLookup[slot]?.originalPcmPath = destPath.absolutePath
            } catch (e: Exception) {
                failed++
            }
        }

        if (failed > 0 && failed == matches.size)
            return "All file copies failed."
        if (failed > 0)
            return "${matches.size - failed} of ${matches.size} tracks imported. $failed file(s) failed."
        return null
    }
}
